### DISCUSSION TITLE
client: select-batching to solve n+1

### DISCUSSION THREAD

**Discussion by mdugue** [geldata/gel#4718] [Q&A]
We are investigating in using EdgeDB to serve a nodejs GraphQL frontend. Unfortunately we haven't found a good & clean solution to circumvent the n+1 Problem. Is there any form of "automated select batching", or "lazy selects" or similar to address this issue?

## n+1

Given a Schema as 

```graphql
type User {
  id: ID!
  posts: [Post]!
}

type Post{
  id: ID! 
}

type Query {
  allUsers: [User]
}
```

we would need to create 2 Reducers, one for the `User`, on for the `Post`, each one would call EdgeDB to retrieve the related data.

Given a Query such as
```graphql
query {
  allUsers {
    id
    posts {
      id
    }
  }
}
```

This would result in one db call per User plus one call per Post for each User. This is commonly known as the n+1 problem. While there are solutions such as https://github.com/graphql/dataloader that can be wrapped around, I was wondering if Edgedb has a native solution to this problem, or is planning to.

**Prisma** follows an interesting approach:

> The Prisma Client dataloader automatically batches `findUnique` queries that ✔ occur in the same tick and ✔ have the same `where` and `include` parameters.

https://www.prisma.io/docs/guides/performance-and-optimization/query-optimization-performance

Is there anything comparable in edgegb? Any hints are highly appreciated 🙏

**Comment by elprans** [comment]
> This would result in one db call per User plus one call per Post for each User. 

In EdgeDB it wouldn't.  EdgeDB always executes queries of any complexity in a single server roundtrip.  Are you observing a different behavior?

**Reply by mdugue** [reply]
wow… Thanks for your reply @elprans . Thats interesting news!. Do I understand correctly, that when the `Query.allUsers` resolver and the `User.posts` each call 
```js
e.select(e.User, () => (e.User['*'])).run(client) // 1x
// respectively 
e.select(e.Post, () => (e.Post['*'])).run(client) // 100x
``` 
this will end up in a single edgedb query internally? Can you share details on how this is working? Is the client batching somwhow and point me to the related documentation? I had the impression that adding `posts` to the `allUsers` query had notable performance impact, but I might be wrong.
